design/automated-provisioning-design.md

```diff
@@ -0,0 +1,147 @@
+Its first alpha/experimental release was in December 2020, and it has remained in alpha/experimental status since then.
+It had significant development through July 2021, at which point the primary maintainer and contributor stepped away from the project.
+Since then, it has been minimally developed, and remains alpha/experimental.
+
+Over the same time, Contour has also made a significant investment in implementing the [Gateway API](https://gateway-api.sigs.k8s.io/).
+There have been several evolutions of Contour’s implementation over that time period.
+One gap we see in Contour’s current implementation is that it lacks support for fully automated provisioning of Contour + Envoy instances, based on a user’s definition of a Gateway custom resource.
+This gap prevents Contour from having a fully conformant Gateway API implementation.
+
+## Terminology
+In this document, I’ll use the terms **static provisioning** and **dynamic provisioning**, borrowed from the [Kubernetes storage world](https://kubernetes.io/docs/concepts/storage/persistent-volumes/#provisioning).
+- **Static provisioning** refers to a user manually deploying a Contour + Envoy instance, plus a Gateway custom resource that describes that Contour + Envoy instance. No automated provisioners are involved; the work is performed by a human. It’s up to that human to ensure that the Gateway spec matches the deployed Contour + Envoy’s configuration.
+- **Dynamic provisioning** refers to the process where a **Gateway provisioner** (a controller running in the cluster) reacts to the creation of a Gateway custom resource by deploying a Contour + Envoy instance that matches the Gateway’s spec. In this process, a human defines a request for load balancing infrastructure, via the Gateway custom resource, and the Gateway provisioner is responsible for fulfilling that request. This term also encompasses the ongoing reconciliation loop between the Gateway custom resource and the underlying Contour + Envoy resources, as well as the deletion process.
+
+It’s possible that we’ll decide on different terms for these concepts going forward (e.g. “managed” vs. “unmanaged”), but this document will stick with the ones defined above.
+
+
+## Goals
+- Create an optional dynamic Gateway provisioner, implemented in `projectcontour/contour` as a subcommand of the `contour` binary and deployed as a separate Deployment, in order to have a conformant Gateway API implementation.
+- Deprecate the Contour operator and associated `Contour` CRD, to focus effort on the Gateway provisioner.
+
+## Design
+The Contour Operator, including the `operator.projectcontour.io/Contour` CRD, will be deprecated.
+We will continue to make the minimal updates required to keep it functioning as-is, and ship new releases as we ship Contour releases, until Gateway API reaches beta (which is expected to be soon, i.e this year) and Contour ships its first release supporting the beta API.
```

Comment:
We probably need an additional caveat in here that we won't complete the process of archiving the operator until there's equivalent functionality available using the gateway-provisioner for things like service annotations etc. There were a number of community PRs to the operator repo for some networking and annotation functionality.

Reply:
:+1: I'll go through the `Contour` CRD to confirm that's reasonable, but yeah I think that's fair. Some of those things will not be included in the MVP, though.